### PR TITLE
41 spherical lensing

### LINKED ISSUE #41: spherical-lensing
Update lensing and work towards spherical lensing

### PR DESCRIPTION
**Pull request by ASKabalan**

### CODE REVIEW

**Comment by EiffL** [comment]
@ASKabalan just a few small comments on good practices:

- remember to add a description of what your PR is doing in the initial post. Here for instance, if you are adding some lensing functionality, you can show a plot or some sort of results of validation tests to give the reviewer a sense of how well the added feature is working.

- it is good practice to keep your PRs atomistic, around well defined scope. Here for instance I see modifications to growth.py, a simplectic integrator, and some lensing utilities. These are pretty unrelated things, it would be better to split into several PRs

File: jaxpm/spherical.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
+from functools import partial
+
+import healpy as hp
+import jax
+import jax.numpy as jnp
+import jax_healpy as jhp
```
Comment (EiffL) [inline jaxpm/spherical.py:6]: Please add the dependency to the pyrpoject.toml

**Review by EiffL — CHANGES_REQUESTED** [review]
- I would suggest changing the approach to bin directly the particles on the spherical maps

- Please remove from this PR things that are not related to spherical lensing, you can put them into a separate PR

- We need to add validation of the spherical lensing maps from somewhere, maybe comparing to glass?

**Comment by EiffL** [comment]
Ignore last message ^^

**Comment by ASKabalan** [comment]
@EiffL 

# Branch History
> remember to add a description of what your PR is doing in the initial post. Here for instance, if you are adding some lensing functionality, you can show a plot or some sort of results of validation tests to give the reviewer a sense of how well the added feature is working.
#
> it is good practice to keep your PRs atomistic, around well defined scope. Here for instance I see modifications to growth.py, a simplectic integrator, and some lensing utilities. These are pretty unrelated things, it would be better to split into several PRs


You are right .. the other functionalities come from other branch since I created this one out of the fix-sharding branch
I can rewrite history and make sure that  it matches that of main .. but we all know what do you think of that :smiling_face_with_tear: 

>I would suggest instead to bin the particles directly onto the spherical maps? that would be a much better approach.

Yeah I agree ..  I think a find a way to do it better in the litterature

> we can see a very weird pattern on these maps that probably has something to do iwth the 3d box

Yes I figured out that my  D_R goes outwards which means that the first projection includes only the corner of the box since it is longer that the box depth

https://github.com/DifferentiableUniverseInitiative/JaxPM/blob/09d82283f5600de3ddc0070081ebd66f2e5a3c48/jaxpm/spherical.py#L33

This explains that wierd circles .. I just need to integrate the D_R backwards

>Please remove from this PR things that are not related to spherical lensing, you can put them into a separate PR

Can I "slightly" rewrite history of this branch then :pray:

**Comment by EiffL** [comment]
let's discuss the caching in a different thread, I can tell you what my logic was.

**Comment by EiffL** [comment]
I'll stop playing with this for now ^^ I let you integrate it with lensing and adding validation tests against glass to make sure we are getting consistent Born raytracing with glass at least

**Comment by ASKabalan** [comment]
I think this works now 

image

check notebook https://github.com/DifferentiableUniverseInitiative/JaxPM/blob/41-spherical-lensing/notebooks/07-SphericalRayTracing.ipynb

**Comment by EiffL** [comment]
So, how is it going :-) ?

**Comment by ASKabalan** [comment]
I am almost there 
Was struggeling with units of jc.background.radial_comoving_distance
it seems to be Mpc/h
But sometimes people divide by cosmo.h to get it in Mpc
for example https://github.com/LSSTDESC/glass-jax/blob/914451146adaba590eacc23422ce443ebe72a5a3/glass/_src/shells.py#L100

I checked with this https://www.astro.ucla.edu/%7Ewright/CosmoCalc.html
`jc.background.radial_comoving_distance(cosmo, a) / cosmo.h` is in Mpc

I will add a plot of CL in the conversation

**Comment by EiffL** [comment]
This is great progress @ASKabalan ! 

Remember to work in incremental atomic PRs, we can move this discussion to the lensing issue, because here you are testing the impact of different resolutions, but that's a different question from testing if your lensing kernel is working.


I would recommend making a first PR where you implement the spherical binning and lensing operation, and include a comparison against glass as a unit test. This test would be independent from the resolution of the simulation, the question it answers is whether you get the right convergence for a given set of density maps.

**Comment by EiffL** [comment]
And that gradients work... Come to think of it, the nearest cell binning we are doing may not be differentiable.... We might have to implement the equivalent of a cic kernel on the spherical mesh... This needs to be checked

File: jaxpm/spherical.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+    npix = jhp.nside2npix(nside)
+
+    @partial(jax.vmap, in_axes=(0, None, None))
+    def interpolate_volume(coords, volume, pixels):
+        voxels = jax.scipy.ndimage.map_coordinates(volume, coords.T, order=1)
+        sums = jnp.bincount(pixels, weights=voxels, length=npix)
+        return sums
+
+    sum_map = interpolate_volume(coords, volume, pixels).sum(axis=0)
```
[the remark this answers is not in the document]
Comment (ASKabalan) [reply jaxpm/spherical.py:149]: to be removed ... it was the Density to Healpix painting which is a double binning as you said
I will remove it

**Comment by EiffL** [comment]
> 1 - Would you say that the CL coming from the spherical lensing is good enough ?
No they are not good enough, there seems to be a systematically high bias at low ell. In terms of ell_max, we also need to go further (see Cyrile's paper)
image

> 2 - For flat sky let's not worry about it ever? because the cl seems very off
I don't know exactly what evaluation you ran but flat sky certainly is working, c.f. https://github.com/EiffL/LPTLensingComparison/blob/main/notebooks/LPTLensingAccuracyTest.ipynb
image

**Comment by ASKabalan** [comment]
I read cyril's paper and it seems to focus on the shear map mainly on modelisation and the E/B purification which not our problem (for now)

Concerning the overestimation of the power at low ellms .. I did some reading and it could be the fact that the limber approximations (which is used by jax cosmo) can fail at low ells for a very low redshift

I want to compare with pyccl with beyond limber approximation
Do you think this is a waste of time
Anyways with the CIC I did and a appropriate nside value I get

image

Do you consider the missmatch at low ells to be very bad ?
Comparing with a non limber approx is a waste of time?
Should I move on and consider the lensing spectra correct ?

**Comment by EiffL** [comment]
At ell of around 100 we want to have exactly (at the subpercent level) the right amplitude.


But again, it's a differnt question, it's a matter of resolution of the simulation and binning schemes, not a question of correctness of the lensing operation.

**Comment by EiffL** [comment]
oh and yeah comparing with non limber is not a priortiy at all :-)

**Comment by ASKabalan** [comment]
> hummmmmm why is 128 better than 256 ^^? quick note also, it might be a bit dangerous to look at just one realization of these PSs, see my notebook here where I make a few realizations and plot the mean and std: https://github.com/EiffL/LPTLensingComparison/blob/main/notebooks/LPTLensingAccuracyTest.ipynb
> 
> It's true that limber can deviate below ell of 50 or so. But we probably wont use these scales anyway. If you look at cryrille's plot, the first data point is around ell of 30 I think

I projected 128 to nside 64 and 256 to nside 128

Should I compare with same nside ?

**Comment by EiffL** [comment]
well i dont know, what we want is that whatever scheme you choose, you should get the mean of a bunch of realisations to match the theory. in the plot above, looks like blue is better than orange in the range of scales where both are valid.

**Comment by ASKabalan** [comment]
I am transforming this into a playground branch to try do the lensing end to end and discuss

Actual dev is split into 3 PR

- An appropriate spherical binning https://github.com/DifferentiableUniverseInitiative/JaxPM/issues/45
- Validating against born (independant of binning)  https://github.com/DifferentiableUniverseInitiative/JaxPM/issues/47
- Validating against theory (once both are done)

**Comment by ASKabalan** [comment]
@EiffL 

I think this PR Is also good to go (I am doing final touches)

But in short maps seems to be good

image


The differences are acceptable

image

And cl ratios between jaxpm and glass tophat is within 10%

image

I am adding a test
